infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServer.java

```diff
@@ -96,6 +96,7 @@
      * storage.
      */
     private final LogUnitServerCache dataCache;
+    @Getter
     private final StreamLog streamLog;
```

Comment:
I think leaking this outside the LogUnitServer object is not clean. Right now LogUnitServer is almost a thin wrapper around StreamLog, but when this abstraction was introduced I was considering striping. One way of implementing striping is to have a StreamLog per stripe. Anyways, this construct is internal to the LogUnitServer and I think we should keep it that way for aforementioned reason.

 What's the cost of not exposing it and making the LogUnutServer directly support your use case?

Reply:
The problem that we have right now is that there is no fast and easy way to figure out the addresses that are present in the unknown transfer range. S/T can fail for one reason or another, and when it restarts, it's important to understand what we need to transfer within a segment range. This would have been less of a problem if we had the concept of a committed range. Then scanning the part of the non-committed log would not be such a big problem. 

However, right now, we actually have to look at the known addresses, and this information is stored within a particular StreamLog's segmentHandle's known addresses. Performing RPCs to the local LogUnitServer just to request known addresses right now is an inefficient process because we have to scan. Moreover, we are not sure if the range scan in question fits fully within one RPC call without timing out, so we have to batch our requests.

If we want to make LogUnitServer support this use case in a more efficient way, we would have to provide an additional data structure for this case specifically, e.g. a bloom filter, bitsets, etc. This data structure would have to be in sync with an internal StreamLog state and providing the set difference in O(1). I am not sure how viable right now is to design something like this.

Considering all that, I do agree with all the contra points, and it seems like exposing a StreamLog introduces more problems than benefits.

Reply:
Fixed